#### Metode i primjena vjestacke inteligencije
#### Laboratorijska vjezba 2
#### Student: Masovic Haris
#### Index: 1689/17993

## 0. Dependencies

In [1]:
import sys

!{sys.executable} -m pip install tabletext


You should consider upgrading via the '/Library/Frameworks/Python.framework/Versions/3.8/bin/python3.8 -m pip install --upgrade pip' command.


## 2. KNN algoritam

### 2.1 Implementacija KNN algoritma

In [6]:
from collections import Counter
from tabletext import to_text
import math

# set True to show logs
SHOW_LOGS = False

# 2.1.9
classification_data = [
    [34, 1], 
    [66, 0], 
    [76, 0], 
    [51, 1], 
    [91, 0], 
    [56, 1], 
    [46, 1], 
    [31, 1], 
    [21, 1],
    [63, 0], 
    [94, 0], 
    [81, 0], 
    [65, 0], 
    [18, 1], 
    [25, 1]
]

# 2.1.12
regression_data = [
    [159.2, 64.32], 
    [189.1, 96.92], 
    [148.8, 51.58], 
    [201.8, 103.49], 
    [191.1, 150.22],
    [169.2, 85.54], 
    [153.2, 45.99], 
    [170.3, 69.43], 
    [199.5, 200.43], 
    [188.7, 97.7], 
    [165.5, 87.2], 
    [134.6, 80.7]
]


def knn(data, query, k, distance_func, choice_func):
    # print(*[type(data), type(query), type(k), type(distance_func), type(choice_func)])
    neighbor_distances_and_indices = []
    
    for index, example in enumerate(data):
        distance = distance_func(example[:-1], query)
        
        neighbor_distances_and_indices.append((distance, index))
    
    sorted_neighbor_distances_and_indices = sorted(neighbor_distances_and_indices)
    
    k_nearest_distances_and_indices = sorted_neighbor_distances_and_indices[:k]
    
    k_nearest_labels = [data[i][1] for distance, i in k_nearest_distances_and_indices]

    return k_nearest_distances_and_indices , choice_func(k_nearest_labels)

def mean(labels):
    return sum(labels) / len(labels)

def mode(labels):
    return Counter(labels).most_common(1)[0][0]

def neka_distance(point1, point2):
    sum_squared_distance = 0
    for i in range(len(point1)):
        sum_squared_distance += math.pow(point1[i] - point2[i], 2)
    return math.sqrt(sum_squared_distance)

def manhattan_distance(point1, point2):
    sum_squared_distance = 0
    for i in range(len(point1)):
        sum_squared_distance += abs(point1[i] - point2[i])
    return math.sqrt(sum_squared_distance)

def chebyshev_distance(point1, point2):
    max_distance = 0
    for i in range(len(point1)):
        abs_difference = abs(point1[i] - point2[i])
        if abs_difference > max_distance:
            max_distance = abs_difference
    return max_distance

def write_results(header, data, queries, k_values, distance_fun, choice_fun):
    if SHOW_LOGS == True:
        print('')
        print(*[header, 'logs:'])
        print('')
    
    header_table = ['Query/K'] + k_values
    table = [header_table]
    
    for item in queries: 
        row = [item]
        for k in k_values:
            results, choice = knn(data, [item], k, distance_fun, choice_fun)
            row.append(str(choice))
            if SHOW_LOGS == True:
                print(*['query:', item, 'k:', k, 'result:', (results, choice)])
        table.append(row)
        
    print('')
    print(*[header, 'results:'])
    print('')
    
    return table

def main():
    # 2.1.10
    classification_query = [33, 50, 80]
    # 2.1.13
    regression_query = [150.4, 173.4, 190.3]
    
    # 2.1.10
    # 2.1.13
    k_values = [1, 3, 5]
    
    c_result = write_results('Euclid distance - Classification', classification_data, classification_query, k_values, neka_distance, mode)
    
    # 2.1.10 - testiranje
    print(to_text(c_result))
    
    r_result = write_results('Euclid distance - Regression', regression_data, regression_query, k_values, neka_distance, mean)
    
    # 2.1.13 - testiranje
    print(to_text(r_result))
    
    # 2.1.15 - manhattan distanca
    manhattan_c_result = write_results('Manhattan distance - Classification', classification_data, classification_query, k_values, manhattan_distance, mode)
    
    print(to_text(manhattan_c_result))
    
    manhattan_r_result = write_results('Manhattan distance - Regression', regression_data, regression_query, k_values, manhattan_distance, mean)
    
    print(to_text(manhattan_r_result))
    
    # 2.1.15 - chebyshev distanca
    chebyshev_c_result = write_results('Chebyshev distance - Classification', classification_data, classification_query, k_values, chebyshev_distance, mode)
    
    print(to_text(chebyshev_c_result))
    
    chebyshev_r_result = write_results('Chebyshev distance - Regression', regression_data, regression_query, k_values, chebyshev_distance, mean)
    
    print(to_text(chebyshev_r_result))
    

if __name__ == '__main__':
    main()


Euclid distance - Classification results:

┌─────────┬───┬───┬───┐
│ Query/K │ 1 │ 3 │ 5 │
├─────────┼───┼───┼───┤
│      33 │ 1 │ 1 │ 1 │
├─────────┼───┼───┼───┤
│      50 │ 1 │ 1 │ 1 │
├─────────┼───┼───┼───┤
│      80 │ 0 │ 0 │ 0 │
└─────────┴───┴───┴───┘

Euclid distance - Regression results:

┌─────────┬────────┬────────────────────┬─────────┐
│ Query/K │      1 │                  3 │       5 │
├─────────┼────────┼────────────────────┼─────────┤
│   150.4 │ 51.58  │ 53.96333333333333  │ 65.958  │
├─────────┼────────┼────────────────────┼─────────┤
│   173.4 │ 69.43  │ 80.72333333333334  │ 80.838  │
├─────────┼────────┼────────────────────┼─────────┤
│   190.3 │ 150.22 │ 114.94666666666666 │ 129.752 │
└─────────┴────────┴────────────────────┴─────────┘

Manhattan distance - Classification results:

┌─────────┬───┬───┬───┐
│ Query/K │ 1 │ 3 │ 5 │
├─────────┼───┼───┼───┤
│      33 │ 1 │ 1 │ 1 │
├─────────┼───┼───┼───┤
│      50 │ 1 │ 1 │ 1 │
├─────────┼───┼───┼───┤
│      80 │ 0 │ 0

#### 2.1.1 Objasnite koji parametri se salju funkciji knn() na ulazu. O kojoj strukturi i vrsti podataka je rijec (skalar, vektor, int, float, itd.) ??

- Parametar `data` predstavlja listu koja sadrzi elemente tipa `[podatak, labela]`, odnosno nase inicijalne podatke navedene u vjezbi, a takoder ukoliko iskorstimo funkciju `type` dobicemo da je to tip liste, odnosno `<class 'list'>`.
- Parametar `query` takodjer predstavlja listu, ovaj parametar predstavlja vrijednost koja je nama od znacaja bilo za regresiju ili klasifikaciju. Na osnovu te vrijednosti ce se racunati udaljenosti elemenata u skupu podataka. Gledajuci funkciju inicijalni postavljenu `neka_distance` vidimo da ovaj tip mora biti lista, jer u protivnom ce se desiti greska, tako da ce se vrijednosti slati kao `[vrijednost]` u kodu.
- Parametar `k` predstavlja neki cijeli broj `<class 'int'>` koji govori koliko najblizih slusjeda uzeti za nasu predefinisanu `query` tacku i na osnovu njih ce se klasificirati ili regresija uraditi.
- Parametar `distance_func` je tip funkcije odnosno `<class 'function'>`. Ova funkcija predstavlja funkciju koja ce se koristiti za racunanje udaljenosti elemenata iz `data` skupa u odnosu na `query` parametar.
- Parametar `choice_func` je tip funkcije odnosno `<class 'function'>`. Ova funkcija se koristi za finalni odabir predikcije za `query` parametar. Za slucaj klasifikacije koristiti ce se `mode` funkcija, dok za slucaj regresije koristiti ce se `mean` funkcija.

#### 2.1.2 Objasnite sta funkcija knn() vraca na izlazu?

Funkcija `knn()` na izlazu vraca dvije vrijednosti:
- `k_nearest_distances_and_indices`, sto predstavlja listu koja se sastoji od objekata tipa `tuple`. Velicina ove liste predstavlja `k`, i to su vrijednosti koje imaju najmanju udaljenost od `query` parametra,
- predikcija za `query` podatak kad je u pitanju klasifikacija, odnosno u slucaju regresije predstavlja kilograme za poslani `query`.

#### 2.1.3 Kako se izracunava distanca u datom primjeru? Identifikujte o kojoj distanci je rijec (na osnovu informacija datim na predavanjima).

- U datom primjeru distanca se izracunava kao kvadratni korijen sume koja se dobije sabiranjem kvadrata razlike izmedju svih tacaka u `data` skupu i `query` tacke. U pitanju je euklidska distanca. 

#### 2.1.4 Koja struktura se koristi za `neighbor_distances_and_indices` ?

- `neighbor_distances_and_indices` predstavlja listu objekata tipa `tuple`.

#### 2.1.5 Kako su podaci organizovani u strukturi `neighbor_distances_and_indices` ?

- Posto je u pitanju lista `tuple`-ova, prvi element objekta predstavlja distancu koju smo dobili racunanjem distance svih ``` distance = distance_func(example[:-1], query) ``` dok drugi element objekta predstavlja index u strukturi (0, 1, 2..).

#### 2.1.6 Koliko elemenata ce imati struktura `neighbor_distances_and indices`?
- Kad je u pitanju klasifikacija, pocetni dataset ima 15 elemenata, shodno time ce i navedena struktura imati toliko, a u slucaju regresije biti ce 12 elemanata.

#### 2.1.7 Koji algoritam se koristi za sortiranje elemenata unutar ove strukture?

- Algoritam za sortiranje koji je realizovan u `sorted` funkciji se naziva [Timsort](https://en.wikipedia.org/wiki/Timsort). Sortiranje je bazirano na prvom elementu tj. distanci i finalni rezultat je defaultni rastuci poredak.

#### 2.1.8 Koja vrijednost ce biti zapisana u varijablu `k_nearest_distances_and_indices` ?

- U ovu varijablu ce biti upisano prvih `k` elemenata iz `sorted_neighbor_distances_and indices` varijable, tj. prvih `k` elemenata koji imaju najmanju euklidsku distancu.

#### 2.1.9 Definisati strukturu podataka classification data, a potom u nju spremiti podatke iz tabele 1. Svaki red u ovoj strukturi ce imati dva elementa, na primjer: [34, 1], [66, 0], itd.

- Uradjeno i u kodu postavljen komentar `# 2.1.9`.

#### 2.1.10 Sekvencijalno definisati sljedece tri vrijednosti: 1) 33, 2) 50 i 3) 80 u varijablu `classification_query`, a potom sekvencijalno izvrsiti testiranje algoritama na izlazu. Za sva tri slucaja, definisati da je K: a) 1, b) 3 i c) 5, kao sto je prikazano u tabeli 3.

- Uradjeno i u kodu postavljeno `# 2.1.10` gdje treba da bude.

#### 2.1.11 Za jedan od prethodnih slucaja (npr. Query=33 i K=3) matematickim putem pokazati kako se izracunava izlazna vrijednost.

In [3]:
print(classification_data)

[[34, 1], [66, 0], [76, 0], [51, 1], [91, 0], [56, 1], [46, 1], [31, 1], [21, 1], [63, 0], [94, 0], [81, 0], [65, 0], [18, 1], [25, 1]]


- Za racunanje cemo koristit euklidsku distancu, a formula koja se koristi je sljedeca:

\begin{equation*}
distanca = \sqrt{\sum_{i=1}^k{(x_i - y_i)^2}}
\end{equation*}

- Shodno time imamo, gledajuci u `classification_data` dataset:
\begin{equation*}
distanca\ (34) = \sqrt{(34 - 33)^2} = \textbf{1}
\end{equation*}
\begin{equation*}
distanca\ (66) = \sqrt{(66 - 33)^2} = 33
\end{equation*}
\begin{equation*}
distanca\ (76) = \sqrt{(76 - 33)^2} = 43
\end{equation*}
\begin{equation*}
distanca\ (51) = 18
\end{equation*}
\begin{equation*}
distanca\ (91) = 58
\end{equation*}
\begin{equation*}
distanca\ (56) = 23
\end{equation*}
\begin{equation*}
distanca\ (46) = 13
\end{equation*}
\begin{equation*}
distanca\ (31) = \textbf{2}
\end{equation*}
\begin{equation*}
distanca\ (21) = 12
\end{equation*}
\begin{equation*}
distanca\ (63) = 30
\end{equation*}
\begin{equation*}
distanca\ (94) = 61
\end{equation*}
\begin{equation*}
distanca\ (81) = 48
\end{equation*}
\begin{equation*}
distanca\ (65) = 32
\end{equation*}
\begin{equation*}
distanca\ (18) = 15
\end{equation*}
\begin{equation*}
distanca\ (25) = \textbf{8}
\end{equation*}


- Vidimo da za `k=3` imamo da su osobe od 34, 33, 25 godina najblize `query=33` vrijednosti. Sve tri osobe imaju labelu 1, tj. sve tri osobe vole jabuku u smoothie-ju shodno time mozemo zakljuciti da i ta osoba sa 33 godine pripada skupini `1` po `mode` funkciji.

#### 2.1.12 Definisati strukturu podataka `regression_data` i u nju spremiti podatke iz tabele 2.

- Uradjeno i u kodu postavljeno `# 2.1.12` gdje treba da bude.

#### 2.1.13 Sekvencijalno definisati sljedece tri 1) 150.4, 2)173.4 i 3) 190.3 vrijednosti u varijablu regression query, a potom sekvencijalno testirati  sta ce algoritam dati na izlazu. Za sva tri slucaja definisati da je K: a) 1, b) 3 i c) 5, kao sto je prikazano u tabeli 4.

- Uradjeno i u kodu postavljeno `# 2.1.13` gdje treba da bude.

#### 2.1.14. Za jedan od prethodnih slucaja (npr. Query=150.4 i K=3) matematickim putem pokazati kako se izracunava izlazna vrijednost.

In [4]:
print(regression_data)

[[159.2, 64.32], [189.1, 96.92], [148.8, 51.58], [201.8, 103.49], [191.1, 150.22], [169.2, 85.54], [153.2, 45.99], [170.3, 69.43], [199.5, 200.43], [188.7, 97.7], [165.5, 87.2], [134.6, 80.7]]


- Za racunanje cemo koristit euklidsku distancu, a formula koja se koristi je sljedeca:

\begin{equation*}
distanca = \sqrt{\sum_{i=1}^k{(x_i - y_i)^2}}
\end{equation*}


- Shodno time imamo, gledajuci u `classification_data` dataset:
\begin{equation*}
distanca\ (159.2) = \sqrt{(159.2 - 150.4)^2} = \textbf{8.799999999999983}
\end{equation*}
\begin{equation*}
distanca\ (189.1) = \sqrt{(189.1 - 150.4)^2} = 38.69999999999999
\end{equation*}
\begin{equation*}
distanca\ (148.8) = \sqrt{(148.8 - 150.4)^2} = \textbf{1.5999999999999943}
\end{equation*}
\begin{equation*}
distanca\ (201.8) = 51.400000000000006
\end{equation*}
\begin{equation*}
distanca\ (191.1) = 40.69999999999999
\end{equation*}
\begin{equation*}
distanca\ (169.2) = 18.799999999999983
\end{equation*}
\begin{equation*}
distanca\ (153.2) = \textbf{2.799999999999983}
\end{equation*}
\begin{equation*}
distanca\ (170.3) = 19.900000000000006
\end{equation*}
\begin{equation*}
distanca\ (199.5) = 49.099999999999994
\end{equation*}
\begin{equation*}
distanca\ (188.7) = 38.29999999999998
\end{equation*}
\begin{equation*}
distanca\ (165.5) = 15.099999999999994
\end{equation*}
\begin{equation*}
distanca\ (134.6) = 15.800000000000011
\end{equation*}


- Za svaku najmanju vrijendnost mozemo procitati visine. Za te visine mozemo uzeti kilaze i shdno time po `mean` funkciji imamo finalno predikciju kilaze za `Query=150.4`:

\begin{equation*}
    \frac{64.32 + 51.58 + 45.99}{3 } = 53.96 kg
\end{equation*}


#### 2.1.15 Analizirati sta se desava za slucaj izlaznih vrijednosti KNN algoritma kada se izvrse sva testiranja nad prethodno spomenutim skupovima podataka s drugom metrikom za izracunavanje udaljenosti izmedu podataka. Da li je algoritam i u ovom slucaju vratio iste rezultate kao i za prethodni slucaj uzete metrike (distance) za izracunavanje udaljenosti? Sta se desava s rezultatima algoritma ako se upotrijebi neka treca metrika za izracunavanje udaljenosti izmedu tacaka? (Napomena: Za svaku uzetu metriku za izracunavanje udaljenosti potrebno je izvrsiti sva potrebna testiranja).

Uveli smo nove dvije funckije distance:
- Manhattan distancu

\begin{equation*}
distanca = \sum_{i=1}^k{\lvert (x_i - y_i) \rvert}
\end{equation*}

- Chebyshev distancu

\begin{equation*}
distanca = \max_{i}^k{\lvert (x_i - y_i) \rvert}
\end{equation*}


Finalno mozemo zakljuciti:
- Bez obzira na koju od navedene 3 metrike (`euclid`, `manhattan`, `chebyshev`), rezultati su ostali isti.
- Naravno rezultati distanci tj. medju-rezultati su se izmijenili (moze se vidjeti ukoliko kazemo `SHOW_LOGS=True`), medjutim finalne vrijednosti su i dalje ostale iste.

#### 2.1.16 Uporedite testove u odnosu na dosad izmjerene vrijednosti. Kakve zakljucke mozete izvesti? Da li algoritam u svim slucajevima vratio iste rezultate? Zasto?

- Mozemo zakljuciti da koristenjem bilo koje od 3 navedene distance, algoritam vrati iste rezultate. To naravno zavisi od samih podataka, a sve 3 distance racunaju udaljenost najblizih elemenata na dobar nacin podjednako.
- U slucaju da je drugaciji dataset, takav da moze uticati na udaljenosti izmedju tacaka po formulama distance, tada bi se mogla desiti drugacija klasifikacija/predikcija, jer naravno su drugaciji susjedi (drugaciji raspored) kod posmatrane `query` tacke odnosno drugaciji je pocetni dataset.